cmd/kube-proxy/app/server_others.go

```diff
@@ -100,60 +103,73 @@ func (s *ProxyServer) platformSetup() error {
 	return nil
 }
 
+// isIPTablesBased checks whether mode is based on iptables rather than nftables
+func isIPTablesBased(mode proxyconfigapi.ProxyMode) bool {
+	return mode == proxyconfigapi.ProxyModeIPTables || mode == proxyconfigapi.ProxyModeIPVS
+}
+
+// getIPTables returns an array of [IPv4, IPv6] utiliptables.Interfaces. If primaryFamily
+// is not v1.IPFamilyUnknown then it will also separately return the interface for just
+// that family.
+func getIPTables(primaryFamily v1.IPFamily) ([2]utiliptables.Interface, utiliptables.Interface) {
+	execer := exec.New()
+
+	// Create iptables handlers for both families. Always ordered as IPv4, IPv6
+	ipt := [2]utiliptables.Interface{
+		utiliptables.New(execer, utiliptables.ProtocolIPv4),
+		utiliptables.New(execer, utiliptables.ProtocolIPv6),
+	}
+
+	var iptInterface utiliptables.Interface
+	if primaryFamily == v1.IPv4Protocol {
+		iptInterface = ipt[0]
+	} else if primaryFamily == v1.IPv6Protocol {
+		iptInterface = ipt[1]
+	}
+
+	return ipt, iptInterface
+}
+
 // platformCheckSupported is called immediately before creating the Proxier, to check
 // what IP families are supported (and whether the configuration is usable at all).
 func (s *ProxyServer) platformCheckSupported() (ipv4Supported, ipv6Supported, dualStackSupported bool, err error) {
-	execer := exec.New()
-	ipt := utiliptables.New(execer, utiliptables.ProtocolIPv4)
-	ipv4Supported = ipt.Present()
-	ipt = utiliptables.New(execer, utiliptables.ProtocolIPv6)
-	ipv6Supported = ipt.Present()
+	if isIPTablesBased(s.Config.Mode) {
+		ipt, _ := getIPTables(v1.IPFamilyUnknown)
+		ipv4Supported = ipt[0].Present()
+		ipv6Supported = ipt[1].Present()
+
+		if !ipv4Supported && !ipv6Supported {
+			err = fmt.Errorf("iptables is not available on this host")
+		} else if !ipv4Supported {
+			klog.InfoS("No iptables support for family", "ipFamily", v1.IPv4Protocol)
+		} else if !ipv6Supported {
+			klog.InfoS("No iptables support for family", "ipFamily", v1.IPv6Protocol)
+		}
+	} else {
+		// Assume support for both families.
+		// FIXME: figure out how to check for kernel IPv6 support using nft
+		ipv4Supported, ipv6Supported = true, true
```

Comment:
don't you need to do at least
```
	nft.path, err = nft.exec.LookPath("nft")
	if err != nil {
		return nil, fmt.Errorf("could not find nftables binary: %w", err)
	}
```

Reply:
(Clarification: in the earliest version of this PR I never actually checked if nft was installed, so if it wasn't, you got a non-obvious error like in https://github.com/kubernetes/kubernetes/pull/121046#issuecomment-1757357455. But I rewrote things so that [`knftables.New()` does sanity checks now](https://github.com/danwinship/knftables/blob/25063956b4f4833424aae4b1020c481876c1e21d/nftables.go#L75-L84), so it will fail at startup now if `nft` is missing or [if it doesn't work for some reason](https://github.com/kubernetes/kubernetes/pull/121046#issuecomment-1765185980).)

Reply:
```
I1024 01:39:40.191302       6 server_others.go:306] "Using nftables Proxier"
I1024 01:39:40.191321       6 server_others.go:511] "Detect-local-mode set to NodeCIDR, but no PodCIDR defined at node for family" ipFamily="IPv6"
I1024 01:39:40.191329       6 server_others.go:520] "Defaulting to no-op detect-local"
I1024 01:39:40.191428       6 proxier.go:222] "Using nftables mark for masquerade" ipFamily="IPv4" mark="0x00004000"
E1024 01:39:40.196975       6 server.go:550] "Error running ProxyServer" err=<
	unable to create proxier: unable to create ipv4 proxier: could not run nftables command: Error: Could not process rule: Operation not supported
	add table ip kube-proxy
	^^^^^^^^^^^^^^^^^^^^^^^^
 >
E1024 01:39:40.197008       6 run.go:74] "command failed" err=<
	unable to create proxier: unable to create ipv4 proxier: could not run nftables command: Error: Could not process rule: Operation not supported
	add table ip kube-proxy
	^^^^^^^^^^^^^^^^^^^^^^^^
```

Reply:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/121046/pull-kubernetes-e2e-gce-cos-alpha-features/1716622550089338880

Reply:
is there resolution on this?

Reply:
No, I believe if you try to run kube-proxy in nftables mode on a machine with IPv6 disabled, it will exit with an error (similar to the one Antonio quoted from the ChromiumOS job that can't run nftables at all) rather than falling back back to single-stack IPv4 like it's supposed to.

But as seen above, there's a FIXME in the code about this.

(If you were asking about the ChromiumOS error, that's also not resolved, but it doesn't matter until we make nftables the default mode.)